returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,201 @@
+            k: (t.sparse_dim.size if t.sparse_dim else t.shape[-1] if len(t.shape) > 0 else 1, t.ndim)
+            for k, t in self._tensor_dict_template.data.items()
+        }
+        self._default_input = "data" if "data" in self.num_outputs else next(iter(self.num_outputs.keys()))
+        self.num_inputs = self.num_outputs[self._default_input][0]
+
+        self.labels = {}
+        for k, t in self._tensor_dict_template.data.items():
+            if t.vocab:
+                self.labels[k] = t.vocab.labels
+            elif t.sparse_dim:  # sparse_dim but not vocab
+                self.labels[k] = list(map(str, range(t.sparse_dim.dimension)))  # dummy labels
+
+    def init_seq_order(
+        self, epoch: Optional[int] = None, seq_list: Optional[List[str]] = None, seq_order: Optional[List[int]] = None
+    ):
+        """
+        :param epoch:
+        :param seq_list:
+        :param seq_order:
+        :return: whether the order changed (True is always safe to return)
+        """
+        super().init_seq_order(epoch=epoch, seq_list=seq_list, seq_order=seq_order)
+
+        if epoch is None and seq_list is None and seq_order is None:
+            self._num_seqs = 0
+            return True
+
+        assert self._dataset is not None
+        self._dataset.init_seq_order(epoch=epoch, seq_list=seq_list, seq_order=seq_order)
+        self._data_iter = enumerate(self._build_mapping_iter())
+        return True
+
+    def _collect_single_seq(self, seq_idx: int) -> Optional[DatasetSeq]:
+        while True:
+            try:
+                loaded_seq_idx, tensor_dict = next(self._data_iter)
+            except StopIteration:
+                return None
+            assert loaded_seq_idx <= seq_idx, "_collect_single_seq must be done monotonically"
+            if loaded_seq_idx != seq_idx:
+                continue
+            seq = DatasetSeq(features={k: t.raw_tensor for k, t in tensor_dict.data.items()}, seq_idx=seq_idx)
+            return seq
+
+    def _build_mapping_iter(self) -> Iterator[TensorDict]:
+        """
+        :return: an iterator applying both the segment level and across-segment transformations on the given dataset
+        """
+        data_iter = self._iterate_dataset()
+        if self._map_seq_stream is None:
+            return data_iter
+        data_iter = self._map_seq_stream(data_iter)
+        assert isinstance(
+            data_iter, Iterator
+        ), f"map_seq_stream must produce an {Iterator.__name__}, but produced {type(data_iter).__name__}"
+        return data_iter
+
+    def _iterate_dataset(self) -> Iterator[TensorDict]:
+        """
+        :return: generator providing data samples in the form of a TensorDict
+        """
+        data_keys = self._dataset.get_data_keys()
+
+        seq_index = 0
+        while self._dataset.is_less_than_num_seqs(seq_index):
+            self._dataset.load_seqs(seq_index, seq_index + 1)
+            tensor_dict = self._tensor_dict_template.copy_template()
```

Comment:
Hm wait, `self._tensor_dict_template` is about the output (maybe we should rename it further to also reflect that, like `self._out_tensor_dict_template`), but here you have this for the input. We should have a separate `self._in_tensor_dict_template`.

Reply:
Ohh, good catch.

Reply:
Or maybe not necessarily `self._in_tensor_dict_template` but you can directly create it here.

Reply:
It's important though that we use the same for the case it is the same (and thus also the same `Dim`s), e.g. when the user did not specify `map_outputs`